File: java/src/main/java/com/alxarafe/app/domain/inventory/entity/HandlingUnit.java
package com.alxarafe.app.domain.inventory.entity;

import com.alxarafe.app.domain.catalogue.valueobject.ItemId;
import com.alxarafe.app.domain.inventory.valueobject.BatchId;
import com.alxarafe.app.domain.inventory.valueobject.HandlingUnitId;
import com.alxarafe.app.domain.inventory.valueobject.HuStatus;
import com.alxarafe.app.domain.inventory.valueobject.Quantity;
import com.alxarafe.app.domain.inventory.valueobject.Sscc;
import com.alxarafe.app.domain.inventory.valueobject.StockQuantId;
import com.alxarafe.app.domain.topology.valueobject.LocationId;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

/**
 * Aggregate Root representing a Handling Unit (e.g., pallet, box, container).
 *
 * Can contain StockQuants or nest other Handling Units (parent-child relationship).
 */
public final class HandlingUnit {

    private final HandlingUnitId id;
    private final Sscc code;
    private LocationId locationId;
    private HandlingUnitId parentHuId;
    private HuStatus status;
    private final List<StockQuant> quants;

    public HandlingUnit(HandlingUnitId id, Sscc code, LocationId locationId,
                        HandlingUnitId parentHuId, HuStatus status) {
        this.id = Objects.requireNonNull(id, "HandlingUnit id cannot be null.");
        this.code = Objects.requireNonNull(code, "HandlingUnit code cannot be null.");
        this.locationId = locationId;
        this.parentHuId = parentHuId;
        this.status = Objects.requireNonNull(status, "HandlingUnit status cannot be null.");
        if (parentHuId != null && id.equals(parentHuId)) {
            throw new IllegalArgumentException("HandlingUnit cannot be its own parent.");
        }
        this.quants = new ArrayList<>();
    }

    public HandlingUnitId id() {
        return id;
    }

    public Sscc code() {
        return code;
    }

    public Optional<LocationId> locationId() {
        return Optional.ofNullable(locationId);
    }

    public Optional<HandlingUnitId> parentHuId() {
        return Optional.ofNullable(parentHuId);
    }

    public HuStatus status() {
        return status;
    }

    public List<StockQuant> quants() {
        return Collections.unmodifiableList(quants);
    }

    public boolean isAvailable() {
        return status == HuStatus.AVAILABLE;
    }

    public void moveToLocation(LocationId locationId) {
        this.locationId = Objects.requireNonNull(locationId, "LocationId cannot be null.");
        this.parentHuId = null;
    }

    public void nestIntoParent(HandlingUnitId parentHuId) {
        Objects.requireNonNull(parentHuId, "ParentHuId cannot be null.");
        if (this.id.equals(parentHuId)) {
            throw new IllegalArgumentException("HandlingUnit cannot be nested into itself.");
        }
        this.parentHuId = parentHuId;
        this.locationId = null;
    }

    public void releaseFromParent(LocationId newLocationId) {
        this.parentHuId = null;
        this.locationId = newLocationId;
    }

    public void updateStatus(HuStatus status) {
        this.status = Objects.requireNonNull(status, "Status cannot be null.");
    }

    /**
     * Adds an item to this Handling Unit.
     */
    public void addQuant(StockQuantId quantId, ItemId itemId, BatchId batchId, Quantity quantity) {
        Objects.requireNonNull(quantId, "StockQuantId cannot be null.");
        Objects.requireNonNull(itemId, "ItemId cannot be null.");
        Objects.requireNonNull(quantity, "Quantity cannot be null.");

        for (StockQuant quant : quants) {
            if (quant.itemId().equals(itemId) && equalsBatchId(quant.batchId().orElse(null), batchId)) {
                quant.add(quantity);
                return;
            }
        }

        quants.add(new StockQuant(quantId, this.id, itemId, batchId, quantity));
    }

    private static boolean equalsBatchId(BatchId a, BatchId b) {
        if (a == null && b == null) {
            return true;
        }
        if (a == null || b == null) {
            return false;
        }
        return a.equals(b);
    }
}

File: php/src/Domain/Inventory/Entity/HandlingUnit.php
<?php

declare(strict_types=1);

namespace Alxarafe\App\Domain\Inventory\Entity;

use Alxarafe\App\Domain\Catalogue\ValueObject\ItemId;
use Alxarafe\App\Domain\Inventory\ValueObject\BatchId;
use Alxarafe\App\Domain\Inventory\ValueObject\HandlingUnitId;
use Alxarafe\App\Domain\Inventory\ValueObject\HuStatus;
use Alxarafe\App\Domain\Inventory\ValueObject\Quantity;
use Alxarafe\App\Domain\Inventory\ValueObject\Sscc;
use Alxarafe\App\Domain\Inventory\ValueObject\StockQuantId;
use Alxarafe\App\Domain\Topology\ValueObject\LocationId;
use InvalidArgumentException;

/**
 * Aggregate Root representing a Handling Unit (e.g., pallet, box, container).
 *
 * Can contain StockQuants or nest other Handling Units (parent-child relationship).
 */
final class HandlingUnit
{
    /** @var StockQuant[] */
    private array $quants = [];

    public function __construct(
        private readonly HandlingUnitId $id,
        private readonly Sscc $code,
        private ?LocationId $locationId,
        private ?HandlingUnitId $parentHuId,
        private HuStatus $status,
    ) {
        if ($parentHuId !== null && $id->equals($parentHuId)) {
            throw new InvalidArgumentException(
                'HandlingUnit cannot be its own parent.'
            );
        }
    }

    public function id(): HandlingUnitId
    {
        return $this->id;
    }

    public function code(): Sscc
    {
        return $this->code;
    }

    public function locationId(): ?LocationId
    {
        return $this->locationId;
    }

    public function parentHuId(): ?HandlingUnitId
    {
        return $this->parentHuId;
    }

    public function status(): HuStatus
    {
        return $this->status;
    }

    /**
     * @return StockQuant[]
     */
    public function quants(): array
    {
        return $this->quants;
    }

    public function isAvailable(): bool
    {
        return $this->status === HuStatus::AVAILABLE;
    }

    public function moveToLocation(LocationId $locationId): void
    {
        $this->locationId = $locationId;
        // When placed inside a physical location directly, it decoupled from any parent HU
        $this->parentHuId = null;
    }

    public function nestIntoParent(HandlingUnitId $parentHuId): void
    {
        if ($this->id->equals($parentHuId)) {
            throw new InvalidArgumentException(
                'HandlingUnit cannot be nested into itself.'
            );
        }
        $this->parentHuId = $parentHuId;
        // When nested, it inherits/resides in the same location as parent,
        // so its direct physical location reference is cleared.
        $this->locationId = null;
    }

    public function releaseFromParent(?LocationId $newLocationId = null): void
    {
        $this->parentHuId = null;
        $this->locationId = $newLocationId;
    }

    public function updateStatus(HuStatus $status): void
    {
        $this->status = $status;
    }

    /**
     * Adds an item to this Handling Unit.
     */
    public function addQuant(StockQuantId $quantId, ItemId $itemId, ?BatchId $batchId, Quantity $quantity): void
    {
        foreach ($this->quants as $quant) {
            if ($quant->itemId()->equals($itemId) && $this->equalsBatchId($quant->batchId(), $batchId)) {
                $quant->add($quantity);
                return;
            }
        }

        $this->quants[] = new StockQuant(
            $quantId,
            $this->id,
            $itemId,
            $batchId,
            $quantity
        );
    }

    private function equalsBatchId(?BatchId $a, ?BatchId $b): bool
    {
        if ($a === null && $b === null) {
            return true;
        }
        if ($a === null || $b === null) {
            return false;
        }
        return $a->equals($b);
    }
}
